pkg/services/cosmosdb/cosmos-utils.go

```diff
@@ -178,3 +183,138 @@ func deleteDatabase(
+
+	err := pollingUntilReadRegionsReady(ctx, resourceGroupName, accountName, databaseAccountClient)
+	if err != nil {
+		return nil, err
+	}
+	return dt, nil
+}
+
+// For sqlAllInOneManager, the real type of `instance.Details` is `*sqlAllInOneInstanceDetails`,
+// so type assertion must be changed. Expect type assertion, this function is totally the same as previous one.
+// Do you have any good idea make the code cleaner?
+func (s *sqlAllInOneManager) waitForReadRegionsReady(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	dt := instance.Details.(*sqlAllInOneInstanceDetails)
+	resourceGroupName := instance.ProvisioningParameters.GetString("resourceGroup")
+	accountName := dt.DatabaseAccountName
+	databaseAccountClient := s.databaseAccountsClient
+
+	err := pollingUntilReadRegionsReady(ctx, resourceGroupName, accountName, databaseAccountClient)
+	if err != nil {
+		return nil, err
+	}
+	return dt, nil
+}
+
+// For now, this method will return on either context is cancelled or every region's state is "succeeded" in seven consecutive check.
+// The reason why we need seven consecutive check is that the read region is created one by one, there is a small gap between
+// the finishment of previous creation and the start of the next creation. By this check, we can detect gaps shorter than 1 mintue,
+// and report success within 70 seconds after completion.
+func pollingUntilReadRegionsReady(
+	ctx context.Context,
+	resourceGroupName string,
+	accountName string,
+	databaseAccountClient cosmosSDK.DatabaseAccountsClient,
+) error {
+	childCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	ticker := time.NewTicker(time.Second * 10)
+	previousSucceededTimes := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			result, err := databaseAccountClient.Get(childCtx, resourceGroupName, accountName)
+			if err != nil {
+				return err
+			}
+			resultJSONBytes, err := json.Marshal(result)
+			if err != nil {
+				return err
+			}
+			resultJSONString := string(resultJSONBytes)
+
+			//Check whether every read location's state is "Succeeded"
+			allSucceed := true
+			readLocations := gjson.Get(resultJSONString, "properties.readLocations")
+			readLocations.ForEach(func(key, value gjson.Result) bool {
+				state := value.Get("provisioningState").String()
+				if state != "Succeeded" {
+					allSucceed = false
+					previousSucceededTimes = 0
+					return false
+				}
+				return true
+			})
+			if allSucceed && previousSucceededTimes >= 7 {
+				return nil
+			} else if allSucceed {
+				previousSucceededTimes++
+			}
+		}
+	}
+}
+
+func validateReadRegions(
+	context string,
+	regions []string,
+) error {
+	for i := range regions {
+		region := regions[i]
+		if !allowedReadRegions[region] {
+			return service.NewValidationError(
+				fmt.Sprintf("%s.allowedReadRegion", context),
+				fmt.Sprintf("given region %s is not allowed", region),
```

Comment:
is there any log utils?
if yes I think this should use the log instead of sprintf?

Reply:
Sprintf is used in their previous implementation, an example [here](https://github.com/Azure/open-service-broker-azure/blob/master/pkg/services/cosmosdb/plan_schemas.go#L162). If change to log utils, their previous implementation should be modified too.

Reply:
we can leave the previous implementation be.
but it's better to do the right thing in our code.
btw: seems we can use the log like in this file:
https://github.com/Azure/open-service-broker-azure/blob/4cc866a67a61c5c518aa1d985f6b975550bae9ea/pkg/services/mysql/common.go

Reply:
I suggest that we temporarily keep this implementation. Here we want to give an error message with enough information, `Sprintf` is suitable for this case. We can investigate some log utils after main features are implemented.